internal/xpkg/layers.go

```diff
@@ -156,3 +159,106 @@ func AnnotateLayers(i v1.Image) (v1.Image, error) {
 
 	return mutate.ConfigFile(img, cfgFile)
 }
+
+// ImageFromFiles creates a v1.Image from arbitrary files on disk.
+// Each top-level directory at `root` is a separate layer.
+// The function performs no interpretation (parsing) of the files.
+func ImageFromFiles(root string) (v1.Image, error) {
+	extManifest := empty.Image
+
+	entries, err := os.ReadDir(root)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		curDir := filepath.Join(root, entry.Name())
+
+		// Since there is an arbitrary directory of mostly small files, we'll
+		// forego streaming in-memory at the expense of some disk I/O with temporary tarballs.
+		tmpFile, err := os.CreateTemp("", "extension-*.tar")
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = tmpFile.Close() }()
+		if err := createTarball(curDir, tmpFile.Name()); err != nil {
+			return nil, err
+		}
+
+		// Create layer from the tarball file
+		layer, err := tarball.LayerFromFile(tmpFile.Name())
+		if err != nil {
+			return nil, err
+		}
+		// Append layer from the dir tarball
+		extManifest, err = mutate.Append(
+			extManifest,
+			mutate.Addendum{
+				Layer: layer,
+				Annotations: map[string]string{
+					AnnotationKey: entry.Name(),
+				},
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return extManifest, nil
+}
+
+func createTarball(in string, out string) error {
+	f, err := os.Create(filepath.Clean(out))
+	if err != nil {
+		return err
+	}
```

Comment:
It looks like in practice you pass this `tmpFile.Name()`, but `tmpFile` is already a file that's open for reading and writing. Why open it again?

https://pkg.go.dev/os#CreateTemp

Reply:
I changed course a bit and decided to build the tarball in-memory instead, mostly inspired by a similar approach used in other tools building Function images.